File: mvp_app/MyProject/app/src/main/java/com/tosmart/xiexuming/myapplication/file/FileUtil.java
package com.tosmart.xiexuming.myapplication.file;

import com.tosmart.xiexuming.myapplication.callback.FileListCallback;
import com.tosmart.xiexuming.myapplication.thread.ApplicationThreadPool;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.concurrent.ThreadPoolExecutor;

/**
 * @author xiexuming
 */
public class FileUtil {
    private static final String TS_FILE_SUFFIX = "ts";

    private static volatile FileUtil instance;
    private HashMap<String, String> mFileMap;
    private ArrayList<String> mFileList;

    private FileUtil() {
    }

    public static FileUtil getInstance() {
        if (instance == null) {
            synchronized (FileUtil.class) {
                instance = new FileUtil();
            }
        }
        return instance;
    }

    public String getFilePath(String fileName) {
        if (mFileMap.containsKey(fileName)) {
            return mFileMap.get(fileName);
        }
        return null;
    }

    public void requestFileList(File file, FileListCallback callback) {
        Runnable runnable = () -> {
            mFileMap = new HashMap<>(0);
            mFileList = new ArrayList<>();
            updateFileList(file);
            callback.displayFileList(mFileList);
        };
        ThreadPoolExecutor threadPool = ApplicationThreadPool.getInstance().getThreadPool();
        threadPool.execute(runnable);
    }

    private void updateFileList(File rootDir) {
        if (!rootDir.isDirectory()) {
            String path = rootDir.getPath();
            String substring = path.substring(path.lastIndexOf(".") + 1);
            if (substring.equals(TS_FILE_SUFFIX)) {
                String fileName = path.substring(path.lastIndexOf("/") + 1);
                mFileList.add(fileName);
                mFileMap.put(fileName, path);
            }
            return;
        }
        File[] files = rootDir.listFiles();
        if (files == null) {
            return;
        }
        for (File file : files) {
            if (file.isDirectory()) {
                updateFileList(file);
            } else {
                String path = file.getPath();
                String substring = path.substring(path.lastIndexOf(".") + 1);
                if (substring.equals(TS_FILE_SUFFIX)) {
                    String fileName = path.substring(path.lastIndexOf("/") + 1);
                    mFileMap.put(fileName, path);
                    mFileList.add(fileName);
                }
            }
        }
    }

}

File: newpaserapp/app/src/main/java/com/example/newpaserapp/start/FileUtil.kt
package com.xxm.tsparseapp.start

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.withContext
import java.io.File

/**
 * @author xxm
 * 2022/7/20
 **/
class FileUtil {
    private var mFileMap: HashMap<String, String>? = null
    private var mFileList: ArrayList<String>? = null


    fun getFilePath(fileName: String): String? {
        return if (mFileMap!!.containsKey(fileName)) {
            mFileMap!![fileName]
        } else null
    }

    suspend fun requestFileList(file: File?): ArrayList<String>? {
        var job = withContext(Dispatchers.IO) {
            async {
                mFileMap = HashMap(0)
                mFileList = ArrayList()
                file?.let { updateFileList(file) }
                mFileList
            }
        }
        return job.await()
    }

    private fun updateFileList(rootDir: File) {
        if (!rootDir.isDirectory) {
            val path = rootDir.path
            val substring = path.substring(path.lastIndexOf(".") + 1)
            if (substring == TS_FILE_SUFFIX) {
                val fileName = path.substring(path.lastIndexOf("/") + 1)
                mFileList!!.add(fileName)
                mFileMap!![fileName] = path
            }
            return
        }
        val files = rootDir.listFiles() ?: return
        for (file in files) {
            if (file.isDirectory) {
                updateFileList(file)
            } else {
                val path = file.path
                val substring = path.substring(path.lastIndexOf(".") + 1)
                if (substring == TS_FILE_SUFFIX) {
                    val fileName = path.substring(path.lastIndexOf("/") + 1)
                    mFileMap!![fileName] = path
                    mFileList!!.add(fileName)
                }
            }
        }
    }

    companion object {
        @Volatile
        private var instance: FileUtil? = null

        private const val TS_FILE_SUFFIX = "ts"

        @JvmStatic
        fun getInstance(): FileUtil? {
            if (instance == null) {
                synchronized(FileUtil::class.java) { instance = FileUtil() }
            }
            return instance
        }
    }
}